cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStore.java

```diff
@@ -173,8 +192,13 @@ public void delete(String name) throws IOException {
     } finally {
       writeLock.unlock();
     }
-    // Attempt to persist the store.
-    flush();
+    // Attempt to persist the store. If that fails then remove the element and rethrow the Exception.
+    try {
+      flush();
+    } catch (IOException ioe) {
+      delete(name);
```

Comment:
Done. Moved flush in the main try loop. Now that we are failing if flush fails we don't need the separate block.